packages/frontend/src/AfgHandling.ts

```diff
@@ -0,0 +1,265 @@
+
+  // Initializes the `ConsensusView` with empty objects.
+  private initialiseConsensusView(
+    consensusInfo: Types.ConsensusInfo,
+    height: Types.BlockNumber,
+    own: Types.Address,
+    other: Types.Address,
+  ) {
+    const index =
+      consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+
+    let consensusView;
+    if (index === -1) {
+      consensusView = {} as Types.ConsensusView;
+    } else {
+      const found = consensusInfo[index];
+      const [, foundView] = found;
+      consensusView = foundView;
+    }
+
+    if (!consensusView[own]) {
+      consensusView[own] = {} as Types.ConsensusState;
+    }
+
+    if (!consensusView[own][other]) {
+      consensusView[own][other] = {} as Types.ConsensusDetail;
+    }
+
+    if (index === -1) {
+      // append at the beginning
+      consensusInfo.unshift([height, consensusView]);
+    } else {
+      consensusInfo[index] = [height, consensusView];
+    }
+  }
+
+  // Fill the block cache back from the `to` number to the last block.
+  // The function `f` is used to evaluate if we should continue backfilling.
+  // `f` returns false when backfilling the cache should be stopped, true to continue.
+  //
+  // Returns block number until which we backfilled.
+  private backfill(
+    consensusInfo: Types.ConsensusInfo,
+    to: Types.BlockNumber,
+    f: Maybe<(i: Types.BlockNumber, consensusView: Types.ConsensusView) => boolean>,
+    own: Types.Address,
+    other: Types.Address,
+  ): Types.BlockNumber {
+    let cont = true;
+    for (const i of consensusInfo) {
+      const [height,] = i;
+      if (height >= to) {
+        continue;
+      }
+
+      this.initialiseConsensusView(consensusInfo, height, own, other);
+
+      const index =
+        consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+      if (index === -1) {
+        cont = false;
+        break;
+      }
+
+      const [, consensusView] = consensusInfo[index];
+
+      cont = f ? f(height, consensusView) : true;
+      if (!cont) {
+        break;
+      }
+    }
+    return to;
+  }
+
+  private getConsensusView(
+    consensusInfo: Types.ConsensusInfo,
+    height: Types.BlockNumber,
+  ): [Types.ConsensusView, number] {
+    const index =
+      consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+    const [, consensusView] = consensusInfo[index];
+    return [consensusView, index];
+  }
+
+  private updateConsensusInfo(
+    height: Types.BlockNumber,
+    addr: Types.Address,
+    voter: Types.Address,
+    obj: object,
+  ) {
+    const consensusInfo = this.getState().consensusInfo;
+    this.initialiseConsensusView(consensusInfo, height, addr, voter);
+
+    const index = consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+    const [, consensusView] = consensusInfo[index];
+
+    for (const o in obj) {
+      if (obj.hasOwnProperty(o)) {
+        consensusView[addr][voter][o] = obj[o];
+      }
+    }
+
+    if (index !== -1) {
```

Comment:
`index` is never `-1` at this point.

Reply:
Ah thanks, great catch! But I think you meant the other branch, which is supposed to append the block data `[height, consensuView]` at the beginning if it's not already in the `consensusInfo`. This is never necessary because of `initialiseConsensusView`, which already does that. I removed this branch now.